### PR TITLE
Start on more principled unification of sizes.

### PR DESCRIPTION
**Pull request by athas**
Only a single case is handled now (binary operators).  This will have to be tediously extended.

### CODE REVIEW

**Comment by athas** [comment]
Also, this does a slightly orthorgonal refactoring to use `Exp` directly instead of `Size` in more places.

File: src/Language/Futhark/TypeChecker/Unify.hs [hunk 1]
```diff
@@ -559,55 +559,46 @@ unifyWith onDims usage = subunify False
...
+  bcs
+  bound
+  nonrigid
+  (AppExp (BinOp (op1, _) t1 (x1, _) (y1, _) _) _)
+  (AppExp (BinOp (op2, _) t2 (x2, _) (y2, _) _) _) = do
+    unifySizes usage bcs bound nonrigid (Var op1 t1 mempty) (Var op2 t2 mempty)
+    unifySizes usage bcs bound nonrigid x1 x2
+    unifySizes usage bcs bound nonrigid y1 y2
+unifySizes _ _ _ _ d1 d2
+  | SizeExpr d1 == SizeExpr d2 = pure ()
```
Comment (athas) [inline src/Language/Futhark/TypeChecker/Unify.hs:577]: This is the case we should eventually get rid of.

**Comment by athas** [comment]
This crazy @melsman-style type-level programming now not only type checks, but also works in the interpreter:

```Futhark
module meta: {
  type eq[n][m]

  val coerce [n][m]'t : eq[n][m] -> [n]t -> [m]t
  val coerce_inner [n][m]'t [k] : eq[n][m] -> [k][n]t -> [k][m]t

  val refl [n] : eq[n][n]
  val comm [n][m] : eq[n][m] -> eq[m][n]
  val trans [n][m][k] : eq[n][m] -> eq[m][k] -> eq[n][k]

  val plus_comm [a][b] : eq[a+b] [b+a]
  val plus_assoc [a][b][c] : eq[(a+b)+c] [a+(b+c)]
  val mult_comm [a][b] : eq[a*b] [b*a]
  val mult_assoc [a][b][c] : eq[(a*b)*c] [a*(b*c)]
} = {
  type eq[n][m] = [0][n][m]()

  def coerce [n][m]'t (_: eq[n][m]) (a: [n]t) = a :> [m]t
  def coerce_inner [n][m]'t [k] (_: eq[n][m]) (a: [k][n]t) = a :> [k][m]t

  def refl = []
  def comm _ = []
  def trans _ _ = []

  def plus_comm = []
  def plus_assoc = []
  def mult_comm = []
  def mult_assoc = []
}

def main [n][m] (xs: [n]i32) (ys: [m]i32) =
  zip (xs ++ ys) (meta.coerce meta.plus_comm (ys ++ xs))
```

Next step is to make it work in the compiler as well.

**Comment by catvayor** [comment]
this implies to solve #1903

**Comment by athas** [comment]
@catvayor For handling complex expressions in momomorphisation, [perhaps this is useful as inspiration](https://github.com/diku-dk/futhark/blob/ee649e098a612eee6d61893277d9c5d105cfe56c/src/Language/Futhark/Interpreter.hs#L154-L201).

**Comment by athas** [comment]
Also, feel free to merge this into your branch whenever you want.  It has already lived longer than I expected it to.

**Comment by catvayor** [comment]
> Also, feel free to merge this into your branch whenever you want. It has already lived longer than I expected it to.

I was keeping it alive until it's goal was done

**Comment by catvayor** [comment]
I tried to make a test of the type-madness program
```futhark
-- input { [1,2] [3,4] } output { [ (1,3), (2,4), (3,1), (4,2) ] }
```
but `futhark test` told me `unexpected '('`, is it voluntary to not allow tuple, or do I wrote it a bad way ?

**Comment by athas** [comment]
Test blocks only support arrays of primitives.  Change the program to `unzip` the result.

**Pull request merged**